### PR TITLE
Build uncommitted changes at runtime

### PR DESCRIPTION
**Pull request by cdonati**
This is meant to address situations where a build hasn't been run on all of the nodes when there are changes on the head node.

### CODE REVIEW

**Comment by cdonati** [comment]
https://ocd.appscale.com:8080/job/Daily%20Build/2224/
Successful Euca-Faststart: https://ocd.appscale.com:8080/job/Euca-Faststart/354/

File: AppController/djinn.rb [hunk 1]
```diff
@@ -4093,6 +4127,38 @@ def rsync_files(dest_node)
...
+    # Run a build on modified directories so that changes will take effect.
+    get_status = 'git -C appscale status'
+    ssh_opts = "-i #{ssh_key} -o StrictHostkeyChecking=no " +
+      '-o NumberOfPasswordPrompts=0'
+    status = `ssh #{ssh_opts} root@#{ip} #{get_status}`
+
+    if status.include?('AppTaskQueue')
+      Djinn.log_info("Building uncommitted taskqueue changes on #{ip}")
+      build_tq = 'pip install --upgrade --no-deps ' +
+        "#{APPSCALE_HOME}/AppTaskQueue[celery_gui]"
```
Comment (menivaitsi) [inline AppController/djinn.rb]: Can additional requirements like celery_gui be a constant somewhere? Dependencies might change over time.

File: AppController/djinn.rb [hunk 2]
```diff
@@ -4093,6 +4127,38 @@ def rsync_files(dest_node)
...
+        Djinn.log_info("Finished building taskqueue on #{ip}")
+      else
+        Djinn.log_error("Unable to build taskqueue on #{ip}")
+      end
+    end
+
+    if status.include?('AppServer_Java')
+      Djinn.log_info("Building uncommitted Java AppServer changes on #{ip}")
+      build = [
+        "unzip -o #{APPSCALE_CACHE_DIR}/appengine-java-sdk-1.8.4.zip " +
```
Comment (menivaitsi) [inline AppController/djinn.rb]: What if the cache doesn't have the appengine-java-sdk for some reason?

Comment (cdonati) [inline AppController/djinn.rb]: Good point. I'll add an initial check for the file over ssh.

**Comment by cdonati** [comment]
https://ocd.appscale.com:8080/job/Daily%20Build/2228/

**Comment by menivaitsi** [comment]
Is there any way we could also run the build script at this point? 
Would it be too much work restarting components that are already running at this point?

This question assumes that we have build file changes.

**Comment by cdonati** [comment]
Running a full build would be possible, but it would take a lot longer. We'd also have to make changes to the build script so that it can be run while the controller is running.

This is run before any relevant components have been started.

**Comment by cdonati** [comment]
https://ocd.appscale.com:8080/job/Daily%20Build/2280/

**Pull request merged**